autoscale/src/main/java/com/sequenceiq/periscope/service/MandatoryScalingAdjustmentService.java

```diff
@@ -0,0 +1,169 @@
+
+            scalingEventSender.sendScaleUpEvent(loadAlert, existingClusterNodeCount, hostFqdnsToInstanceId.size(), hostIdsWithServicesHealthy.size(),
+                    targetScaleUpCount);
+            LOGGER.info("Triggered mandatory adjustment ScaleUp for Cluster '{}', NodeCount '{}', HostGroup '{}'", cluster.getStackCrn(),
+                    targetScaleUpCount, policyHostGroup);
+        } else if (downscalable(scalingAdjustmentParameters.getMandatoryDownscaleAdjustment())) {
+            List<String> hostsToDecommission = collectDownscaleRecommendations(hostFqdnsToInstanceId, emptyList(),
+                    pollingUserCrn, stackV4Response, scalingAdjustmentParameters.getMandatoryDownscaleAdjustment());
+
+            scalingEventSender.sendScaleDownEvent(loadAlert, hostFqdnsToInstanceId.size(), hostsToDecommission, hostIdsWithServicesHealthy.size(), REGULAR);
+            LOGGER.info("Triggered mandatory adjustment ScaleDown for Cluster '{}', HostsToDecommission '{}', HostGroup '{}'",
+                    cluster.getStackCrn(), hostsToDecommission, policyHostGroup);
+        }
+    }
+
+    private boolean upscalable(Optional<Integer> adjustmentParam) {
+        return adjustmentParam.isPresent() && isCoolDownTimeElapsed(cluster.getStackCrn(), "scaled-up",
+                loadAlertConfiguration.getScaleUpCoolDownMillis(), cluster.getLastScalingActivity());
+    }
+
+    private boolean downscalable(Optional<Integer> adjustmentParam) {
+        return adjustmentParam.isPresent() && isCoolDownTimeElapsed(cluster.getStackCrn(), "scaled-down",
+                loadAlertConfiguration.getScaleDownCoolDownMillis(), cluster.getLastScalingActivity());
+    }
+
+    private List<String> collectDownscaleRecommendations(Map<String, String> hostFqdnsToInstanceId, List<String> stoppedHostIds,
+            String pollingUserCrn, StackV4Response stackResponse, Optional<Integer> mandatoryDownscaleCount) {
+        try {
+            YarnScalingServiceV1Response yarnResponse = yarnMetricsClient.getYarnMetricsForCluster(cluster, stackResponse, policyHostGroup,
+                    pollingUserCrn, mandatoryDownscaleCount);
+
+            int maxAllowedDownscale = Math.max(0, hostFqdnsToInstanceId.size() - loadAlertConfiguration.getMinResourceValue());
+            int allowedDownscale = IntStream.of(mandatoryDownscaleCount.get(), maxAllowedDownscale,
+                            loadAlertConfiguration.getMaxScaleDownStepSize()).min().getAsInt();
+
+            List<String> yarnRecommendedDecommissionHosts = yarnResponseUtils.getYarnRecommendedDecommissionHostsForHostGroup(yarnResponse,
+                    hostFqdnsToInstanceId, maxAllowedDownscale, mandatoryDownscaleCount, loadAlertConfiguration.getMaxScaleDownStepSize());
+
+            return Stream.concat(stoppedHostIds.stream(), yarnRecommendedDecommissionHosts.stream())
```

Comment:
Any need to de-dupe this list ? Typically this will not have dupes. In case they exist - would the actual APIs which are sent this list handle it properly.

Reply:
This list is initialised into a HashSet when sent to cloudbreak which will remove duplicates. But since YARN will not recommend stoppedNodes as downscale candidates, it should not have any duplicate values